src/node-explorer-provider.ts

```diff
@@ -327,6 +214,127 @@ export class NodeExplorerProvider
     }
   }
 
+  refreshAll() {
+    this._onDidChangeTreeData.fire(undefined);
+  }
+
+  public async handleDrop(
+    target: FileExplorer | undefined,
+    sources: vscode.DataTransfer,
+    _: vscode.CancellationToken
+  ): Promise<void> {
+    if (!target) {
+      return;
+    }
+
+    let transferItem = sources.get('application/vnd.code.tree.tsFileEntry');
+    if (transferItem && transferItem.value) {
+      const source: FileExplorer = transferItem.value[0];
+      return this.moveFile(source.uri, target);
+    }
+
+    transferItem = sources.get('text/uri-list');
+    if (transferItem && transferItem.value) {
+      transferItem.value.split('\r\n').forEach((uri: string) => {
+        return this.transferFile(vscode.Uri.parse(uri), target);
+      });
+    }
+
+    sources.forEach(({ value }, mimeType) => {
+      console.log('mimeType', mimeType);
+
+      switch (mimeType) {
+        case 'application/vnd.code.tree.tsFileEntry':
+          return this.transferFile(value.uri, target);
+
+        // From the @types/vscode package: text/uri-list
+        //   A string with `toString()`ed Uris separated by `\r\n`.
+        case 'text/uri-list':
+          value.split('\r\n').forEach((uri: string) => {
+            return this.transferFile(vscode.Uri.parse(uri), target);
+          });
+          return;
+
+        default:
+          return;
+      }
+    });
+  }
+
+  /**
+   * Move a file from within a node or across nodes.
+   */
+  async moveFile(source: vscode.Uri, target: FileExplorer) {
+    const { address: sourceAddr } = parseTsUri(source);
+    const { address: targetAddr } = parseTsUri(target.uri);
+
+    if (sourceAddr !== targetAddr) {
+      // move across nodes
+      const contents = await vscode.workspace.fs.readFile(source);
+      const fileName = path.basename(source.toString());
+      const targetDir = target.getDirectory(fileName);
+
+      await this.fsProvider.writeFile(targetDir, contents, {
+        create: true,
+        overwrite: false,
+      });
+
+      this.refreshAll();
```

Comment:
I think this can also be optimized like transferFile below to update the target only if it's a directory (until/if we have parent linked lists)

Reply:
Optimized where I could. With moving, we don't have a handle to the FileExplorer object it's dropped onto. When we start maintaining a tree to improve refreshes we will want to be able to look them up via Uri for this reason.

Reply:
I created an issue to track:  https://github.com/tailscale-dev/vscode-tailscale/issues/182